xla/service/gpu/llvm_gpu_backend/amdgpu_backend.cc

```diff
@@ -338,60 +339,45 @@ absl::StatusOr<std::vector<uint8_t>> EmitModuleToHsaco(
     }
   } else {
     // Locate lld.
-    std::string lld_path;
     llvm::SmallVector<std::string, 3> lld_paths;
 
     if (const char* llvm_path = std::getenv("LLVM_PATH")) {
-      lld_paths.push_back(tsl::io::JoinPath(std::getenv("LLVM_PATH"), "bin"));
+      lld_paths.push_back(tsl::io::JoinPath(llvm_path, "bin"));
     }
     lld_paths.push_back(tsl::io::JoinPath(tsl::RocmRoot(), "llvm/bin"));
 
     // push LLD path from JAX plugin if set
-    auto jax_paths = getJaxPluginPaths();
-    if (!jax_paths.lld_path.empty()) {
-      lld_paths.push_back(jax_paths.lld_path);
+    if (const auto& jpaths = getJaxPluginPaths(); !jpaths.lld_path.empty()) {
+      lld_paths.push_back(jpaths.lld_path);
     }
 
-    auto lld_program = llvm::sys::findProgramByName("ld.lld", {lld_path});
+    auto lld_program = llvm::sys::findProgramByName(
+      "ld.lld", llvm::to_vector_of<llvm::StringRef>(lld_paths));
     if (!lld_program) {
       return xla::Internal("unable to find ld.lld in PATH: %s",
-                           lld_program.getError().message());
+                         lld_program.getError().message());
     }
-    std::vector<llvm::StringRef> lld_args{
-        llvm_ir::AsStringRef("ld.lld"),
-        llvm_ir::AsStringRef("-flavor"),
-        llvm_ir::AsStringRef("gnu"),
-        llvm_ir::AsStringRef("-shared"),
-        llvm_ir::AsStringRef("--no-undefined"),
-        llvm_ir::AsStringRef(isabin_path),
-        llvm_ir::AsStringRef("-o"),
-        llvm_ir::AsStringRef(hsaco_path),
+    std::initializer_list<llvm::StringRef> lld_args{
+        "ld.lld", "-flavor", "gnu", "-shared",
+        "--no-undefined", isabin_path, "-o", hsaco_path,
     };
 
     std::string error_message;
     int lld_result =
-        llvm::sys::ExecuteAndWait(*lld_program, llvm_ir::AsArrayRef(lld_args),
+        llvm::sys::ExecuteAndWait(*lld_program, lld_args,
                                   std::nullopt, {}, 0, 0, &error_message);
     if (lld_result) {
       return xla::Internal("ld.lld execute fail: %s, error code %d",
                            error_message, lld_result);
     }
-  }
+  } // use_inprocess_lld
 
-  // Read HSACO.
-  std::ifstream hsaco_file(hsaco_path, std::ios::binary | std::ios::ate);
-  std::ifstream::pos_type hsaco_file_size = hsaco_file.tellg();
-
-  std::vector<uint8_t> hsaco(hsaco_file_size);
-  hsaco_file.seekg(0, std::ios::beg);
-  hsaco_file.read(reinterpret_cast<char*>(hsaco.data()), hsaco_file_size);
-  hsaco_file.close();
-  if (!keep_tempfiles) {
-    remove(ir_path.c_str());
-    remove(isabin_path.c_str());
+  if (!HsacoCache::i().keep_temp_files()) {
+    unlink(ir_path.c_str());
+    unlink(isabin_path.c_str());
+    unlink(ir_opt_path.c_str());
```

Comment:
may I ask where these `unlink()` implementation?

Reply:
thanks! i'm wondering why we not use `std::remove()` which is general used and changed to `unlink()` which IIUC is POSIX, and just rm the name and possibly to rm the file if no link for it anymore.